// @okta/okta-angular/esm5/src/okta/models/token-manager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function AccessToken() { }
if (false) {
    /** @type {?} */
    AccessToken.prototype.accessToken;
}
/**
 * @record
 */
export function IDToken() { }
if (false) {
    /** @type {?} */
    IDToken.prototype.idToken;
    /** @type {?} */
    IDToken.prototype.claims;
}
/**
 * @record
 */
export function TokenManager() { }
if (false) {
    /**
     * @param {?} key
     * @return {?}
     */
    TokenManager.prototype.get = function (key) { };
    /**
     * @param {?} key
     * @param {?} token
     * @return {?}
     */
    TokenManager.prototype.add = function (key, token) { };
    /**
     * @param {?} event
     * @param {?} handler
     * @return {?}
     */
    TokenManager.prototype.on = function (event, handler) { };
    /**
     * @param {?} event
     * @param {?} handler
     * @return {?}
     */
    TokenManager.prototype.off = function (event, handler) { };
}
