// @okta/okta-angular/esm2015/src/okta/services/okta.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * Copyright (c) 2017-Present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */
import { Inject, Injectable, Injector } from '@angular/core';
import { assertIssuer, assertClientId, assertRedirectUri, } from '@okta/configuration-validation';
import { OKTA_CONFIG } from '../models/okta.config';
import packageInfo from '../packageInfo';
/**
 * Import the okta-auth-js library
 */
import OktaAuth from '@okta/okta-auth-js';
import { Observable } from 'rxjs';
export class OktaAuthService {
    /**
     * @param {?} config
     * @param {?} injector
     */
    constructor(config, injector) {
        this.injector = injector;
        this.observers = [];
        /**
         * Cache the auth config.
         */
        this.config = Object.assign({}, config);
        this.config.scopes = this.config.scopes || ['openid', 'email'];
        /**
         * Scrub scopes to ensure 'openid' is included
         */
        this.scrubScopes(this.config.scopes);
        // Assert Configuration
        assertIssuer(this.config.issuer, this.config.testing);
        assertClientId(this.config.clientId);
        assertRedirectUri(this.config.redirectUri);
        this.oktaAuth = new OktaAuth(this.config);
        this.oktaAuth.userAgent = `${packageInfo.name}/${packageInfo.version} ${this.oktaAuth.userAgent}`;
        this.$authenticationState = new Observable((observer) => { this.observers.push(observer); });
    }
    /**
     * @param {?=} fromUri
     * @param {?=} additionalParams
     * @return {?}
     */
    login(fromUri, additionalParams) {
        this.setFromUri(fromUri);
        /** @type {?} */
        const onAuthRequired = this.config.onAuthRequired;
        if (onAuthRequired) {
            return onAuthRequired(this, this.injector);
        }
        return this.loginRedirect(undefined, additionalParams);
    }
    /**
     * @return {?}
     */
    getTokenManager() {
        return this.oktaAuth.tokenManager;
    }
    /**
     * Checks if there is an access token OR an id token
     * A custom method may be provided on config to override this logic
     * @return {?}
     */
    isAuthenticated() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // Support a user-provided method to check authentication
            if (this.config.isAuthenticated) {
                return (this.config.isAuthenticated)(this);
            }
            /** @type {?} */
            const accessToken = yield this.getAccessToken();
            /** @type {?} */
            const idToken = yield this.getIdToken();
            return !!(accessToken || idToken);
        });
    }
    /**
     * @param {?} state
     * @return {?}
     */
    emitAuthenticationState(state) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.observers.forEach(observer => observer.next(state));
        });
    }
    /**
     * Returns the current accessToken in the tokenManager.
     * @return {?}
     */
    getAccessToken() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                /** @type {?} */
                const accessToken = (/** @type {?} */ (yield this.oktaAuth.tokenManager.get('accessToken')));
                return accessToken.accessToken;
            }
            catch (err) {
                // The user no longer has an existing SSO session in the browser.
                // (OIDC error `login_required`)
                // Ask the user to authenticate again.
                return undefined;
            }
        });
    }
    /**
     * Returns the current idToken in the tokenManager.
     * @return {?}
     */
    getIdToken() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                /** @type {?} */
                const idToken = (/** @type {?} */ (yield this.oktaAuth.tokenManager.get('idToken')));
                return idToken.idToken;
            }
            catch (err) {
                // The user no longer has an existing SSO session in the browser.
                // (OIDC error `login_required`)
                // Ask the user to authenticate again.
                return undefined;
            }
        });
    }
    /**
     * Returns user claims from the /userinfo endpoint.
     * @return {?}
     */
    getUser() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.oktaAuth.token.getUserInfo();
        });
    }
    /**
     * Returns the configuration object used.
     * @return {?}
     */
    getOktaConfig() {
        return this.config;
    }
    /**
     * Launches the login redirect.
     * @param {?=} fromUri
     * @param {?=} additionalParams
     * @return {?}
     */
    loginRedirect(fromUri, additionalParams) {
        if (fromUri) {
            this.setFromUri(fromUri);
        }
        /** @type {?} */
        const params = Object.assign({
            scopes: this.config.scopes,
            responseType: this.config.responseType
        }, additionalParams);
        return this.oktaAuth.token.getWithRedirect(params); // can throw
    }
    /**
     * Stores the intended path to redirect after successful login.
     * @param {?=} fromUri
     * @return {?}
     */
    setFromUri(fromUri) {
        // Use current location if fromUri was not passed
        fromUri = fromUri || window.location.href;
        // If a relative path was passed, convert to absolute URI
        if (fromUri.charAt(0) === '/') {
            fromUri = window.location.origin + fromUri;
        }
        sessionStorage.setItem('referrerPath', fromUri);
    }
    /**
     * Returns the referrer path from localStorage or app root.
     * @return {?}
     */
    getFromUri() {
        /** @type {?} */
        const fromUri = sessionStorage.getItem('referrerPath') || window.location.origin;
        sessionStorage.removeItem('referrerPath');
        return fromUri;
    }
    /**
     * Parses the tokens from the callback URL.
     * @return {?}
     */
    handleAuthentication() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            const res = yield this.oktaAuth.token.parseFromUrl();
            /** @type {?} */
            const tokens = res.tokens;
            if (tokens.accessToken) {
                this.oktaAuth.tokenManager.add('accessToken', (/** @type {?} */ (tokens.accessToken)));
            }
            if (tokens.idToken) {
                this.oktaAuth.tokenManager.add('idToken', (/** @type {?} */ (tokens.idToken)));
            }
            if (yield this.isAuthenticated()) {
                this.emitAuthenticationState(true);
            }
        });
    }
    /**
     * Clears the user session in Okta and removes
     * tokens stored in the tokenManager.
     * @param {?=} options
     * @return {?}
     */
    logout(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            let redirectUri = null;
            options = options || {};
            if (typeof options === 'string') {
                redirectUri = options;
                // If a relative path was passed, convert to absolute URI
                if (redirectUri.charAt(0) === '/') {
                    redirectUri = window.location.origin + redirectUri;
                }
                options = {
                    postLogoutRedirectUri: redirectUri
                };
            }
            yield this.oktaAuth.signOut(options);
            this.emitAuthenticationState(false);
        });
    }
    /**
     * Scrub scopes to ensure 'openid' is included
     * @param {?} scopes
     * @return {?}
     */
    scrubScopes(scopes) {
        if (scopes.indexOf('openid') >= 0) {
            return;
        }
        scopes.unshift('openid');
    }
}
OktaAuthService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
OktaAuthService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [OKTA_CONFIG,] }] },
    { type: Injector }
];
if (false) {
    /** @type {?} */
    OktaAuthService.prototype.oktaAuth;
    /** @type {?} */
    OktaAuthService.prototype.config;
    /** @type {?} */
    OktaAuthService.prototype.observers;
    /** @type {?} */
    OktaAuthService.prototype.$authenticationState;
    /** @type {?} */
    OktaAuthService.prototype.injector;
}
